src/main/java/com/google/devtools/build/lib/metrics/MetricsCollector.java

```diff
@@ -322,17 +353,50 @@ private TimingMetrics finishTimingMetrics() {
     return timingMetrics.build();
   }
 
+  static class DurationAdder {
+
+    final LongAdder duration;
+    final AtomicBoolean valid;
+
+    public DurationAdder() {
+      valid = new AtomicBoolean(true);
+      duration = new LongAdder();
+    }
+
+    public void add(Duration duration) {
+      if (this.valid.get())
+        this.duration.add(duration.toMillis());
+    }
+
+    public void invalidate() {
+      this.valid.set(false);
+    }
+
+    public boolean isEmpty() {
```

Comment:
Please rename this. Or make `get()` return an Optional. If you do an Optional, the new BuildActionData parts could be nicely expressed with lambdas.